### ISSUE TITLE
Reusing freed memory?

### ISSUE THREAD

**Issue by quesnel** [Nelarius/imnodes#28]
First, thank you for this library!

However, IMHO, The `find_or_create_new` function in `ObjectPool` can lead to a memory error. Indeed, this function allocates memory to grow memory for internal container objects (Using the  `ImVector::resize` function). But many structures store `NodeData`, `PinData` or `NodeData` pointer directly.

In my tests if the number of `PinData` grows (from 8, to 12, then 16), a random error occured in `ImVec2 get_screen_space_pin_coordinates(const PinData& pin)` with random data in `PinData`. 
To quickly fix it, I add the following `reserve` in `EditorContext` constructor but this is not a real solution.

````
EditorContext()
    : nodes(), pins(), links(), panning(0.f, 0.f), link_dragged(),
      box_selector()
{
  nodes.pool.reserve(1024);
  nodes.in_use.reserve(1024);
  nodes.free_list.reserve(1024);
  nodes.id_map.Data.reserve(1024);

  pins.pool.reserve(1024);
  pins.in_use.reserve(1024);
  pins.free_list.reserve(1024);
  pins.id_map.Data.reserve(1024);

  links.pool.reserve(1024);
  links.in_use.reserve(1024);
  links.free_list.reserve(1024);
  links.id_map.Data.reserve(1024);
}
````

**Comment by Nelarius** [comment]
Hi, and thanks for the kind words!

You're absolutely right, the library is referencing freed memory 🤦‍♂The culprit is in these recent commits: `194e949`, `f3c4051`, `7471bbe`, where I refactored the code to use more direct pointers instead of indexes. I just got really lucky with my memory in my local tests and it didn't crash, so I didn't think too hard about it. Classic.

For now, you are better off using the latest tagged version: https://github.com/Nelarius/imnodes/releases/tag/v0.2 It still has the old index indirection and should work better on your machine! I will have to revert some of the changes in master and sort this out.
